plugin/pkg/scheduler/schedulercache/cache.go

```
@@ -0,0 +1,237 @@
+	cache := newSchedulerCache(ttl, cleanAssumedPeriod, stop)
+	cache.run()
+	return cache
+}
+
+type schedulerCache struct {
+	stop   chan struct{}
+	ttl    time.Duration
+	period time.Duration
+
+	// This mutex guards all fields within this cache struct.
+	mu          sync.Mutex
+	assumedPods map[string]assumedPod
+	podStates   map[string]podState
+	nodes       map[string]*NodeInfo
+}
+
+type podState int
+
+const (
+	podAsummed podState = iota + 1
+	podAdded   podState = iota + 1
+)
+
+type assumedPod struct {
+	pod      *api.Pod
+	deadline time.Time
+}
+
+func newSchedulerCache(ttl, period time.Duration, stop chan struct{}) *schedulerCache {
+	return &schedulerCache{
+		ttl:    ttl,
+		period: period,
+		stop:   stop,
+
+		nodes:       make(map[string]*NodeInfo),
+		podStates:   make(map[string]podState),
+		assumedPods: make(map[string]assumedPod),
+	}
+}
+
+func (cache *schedulerCache) GetNodeNameToInfoMap() (map[string]*NodeInfo, error) {
+	nodeNameToInfo := make(map[string]*NodeInfo)
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+	for name, info := range cache.nodes {
+		nodeNameToInfo[name] = info.Clone()
+	}
+	return nodeNameToInfo, nil
+}
+
+func (cache *schedulerCache) List(selector labels.Selector) ([]*api.Pod, error) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	var pods []*api.Pod
+	for _, info := range cache.nodes {
+		for _, pod := range info.pods {
+			if selector.Matches(labels.Set(pod.Labels)) {
+				pods = append(pods, pod)
+			}
+		}
+	}
+	return pods, nil
+}
+
+func (cache *schedulerCache) AssumePodIfBindSucceed(pod *api.Pod, bind func() bool) error {
+	return cache.assumePodIfBindSucceed(pod, bind, time.Now())
+}
+
+// assumePodScheduled exists for making test deterministic by taking time as input argument.
+func (cache *schedulerCache) assumePodIfBindSucceed(pod *api.Pod, bind func() bool, now time.Time) error {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	if !bind() {
+		return nil
+	}
+
+	key := mustGetPodKey(pod)
```

Moving discussion from parent PR here:

@wojtekt
"I'm against having "mustGetPodKey" function. We shouldn't have any "panic" in the code in places where it depends on the data (there can be some data-corruption it shouldn't blow up our components).
It should be:

key, err := getPodKey(pod)
if err != nil {
return err
}

And you can move it outside lock BTW."

@hongchaodeng 
"To return error is not intended, because this error case is not transient error (e.g. data corruption). The pod must have a unique key, e.g. a name.

Currently, Pod.Name is unique within Pod.Namespace, and we get "Namespace/Name" as key by using MetaNamespaceKeyFunc. If some changes breaks this compatibility, it is very dangerous and we shouldn't swallow the error.

We can extract namespace and name from api.Pod directly. But depending on MetaNamespaceKeyFunc seems more maintainable.

Not quite sure about this part. Just let you know that we shouldn't let such error not handled."